src/groth16/verifier/batch.rs

```diff
@@ -167,4 +173,117 @@ where
             Err(VerificationError::InvalidProof)
         }
     }
+
+    /// Perform batch verification with a particular `VerifyingKey`, returning
+    /// `Ok(())` if all proofs were verified and `VerificationError` otherwise.
+    ///
+    /// This performs the bulk of internal arithmetic over the global rayon
+    /// threadpool.
+    #[cfg(feature = "multicore")]
+    #[allow(non_snake_case)]
+    pub fn verify_multicore(self, vk: &VerifyingKey<E>) -> Result<(), VerificationError> {
+        if self
+            .items
+            .iter()
+            .any(|Item { inputs, .. }| inputs.len() + 1 != vk.ic.len())
+        {
+            return Err(VerificationError::InvalidVerifyingKey);
+        }
+
+        struct Accumulator<E: MultiMillerLoop> {
+            gammas: Vec<E::Fr>,
+            delta: E::G1,
+            y: E::Fr,
+            ml_result: Option<E::Result>,
+        }
+
+        impl<E: MultiMillerLoop> Accumulator<E> {
+            fn new(ic_len: usize) -> Self {
+                Accumulator {
+                    gammas: vec![E::Fr::zero(); ic_len],
+                    delta: E::G1::identity(),
+                    y: E::Fr::zero(),
+                    ml_result: None,
+                }
+            }
+        }
+
+        let ic_len = vk.ic.len();
+
+        let acc = self
+            .items
+            // This chunk size was obtained heuristically.
+            .par_chunks(8)
+            .map(|items| {
+                let mut acc = Accumulator::<E>::new(ic_len);
+                let mut ml_terms: Vec<(E::G1Affine, E::G2Prepared)> = vec![];
+                let z = loop {
+                    let z = E::Fr::random(&mut OsRng);
+                    if !z.is_zero_vartime() {
+                        break z;
+                    }
+                };
+                let mut cur_z = z;
+                for Item { proof, inputs } in items {
+                    acc.gammas[0] += &cur_z;
+                    for (a_i, acc_gamma_i) in
+                        Iterator::zip(inputs.iter(), acc.gammas.iter_mut().skip(1))
+                    {
+                        *acc_gamma_i += &(cur_z * a_i);
+                    }
+                    acc.delta += proof.c * cur_z;
+                    acc.y += &cur_z;
+                    ml_terms.push(((proof.a * cur_z).into(), (-proof.b).into()));
+
+                    cur_z *= z;
```

Comment:
Proofs within each chunk are now separated by powers of a common challenge, rather than sampling independent challenges per proof. [Appendix B.2 of the Zcash protocol spec](https://zips.z.cash/protocol/protocol.pdf#grothbatchverify) states:

> The $z_j$ values **MUST** be chosen independently of the _proof batch entries_.

I think this is satisfied, because `z` is sampled independently of all proofs in the chunk, so its powers will be uncorrelated to the chunk batch entries.